### PR TITLE
Methods for creating projects from url

### PR DESCRIPTION
**Pull request by lachmanfrantisek**
- Add method for creating projects from url.
- Get project from url.

TODO:

- [x] add a way to use existing service classes
- [x] add a way to update mapping by user
- [ ] consider implementing API checks if we cannot find a match in mapping
- [x] tests
    - [x] make the `github_repo` lazy for easier testing
    - [x] tests for geting project/service class from url
    - [x] tests for parsing

----

How to test?

```python
from ogr import get_service_class

get_service_class(url="github.com/packit-service/ogr")
get_service_class(url="src.fedoraproject.org/rpms/python-ogr")
```
```python
from ogr import get_project
from ogr import GithubService, PagureService

get_project(
    url="src.fedoraproject.org/rpms/python-ogr",
    custom_instances=[
        GithubService(token="asda"),
        PagureService(instance_url="pagure.io", token="qwe"),
        PagureService(instance_url="src.fedoraproject.org", token="asdaasda"),
    ],
)

```

### CODE REVIEW

**Comment by lachmanfrantisek** [comment]
I've split the `ogr/utils.py` because of the import cycles...

**Comment by TomasTomecek** [comment]
Just FYI, it would be awesome if we got rid of this file https://github.com/packit-service/packit/blob/master/packit/ogr_services.py

**Comment by lachmanfrantisek** [comment]
> Just FYI, it would be awesome if we got rid of this file https://github.com/packit-service/packit/blob/master/packit/ogr_services.py

You mean the [get_github_project method](https://github.com/packit-service/packit/blob/master/packit/ogr_services.py#L95)? Yes, the next step would be to use this in packit...

**Comment by lachmanfrantisek** [comment]
Job retrigered:

https://ci.centos.org/job/ogr-pr/153/console

> [Pipeline] End of Pipeline
> Finished: SUCCESS

File: ogr/parsing.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+def get_username_from_git_url(url: str) -> Optional[str]:
+    """http://github.com/foo/bar.git -> foo"""
+    repo_url = parse_git_repo(url)
+    if repo_url:
+        return repo_url.username
+    return None
+
+
+def get_reponame_from_git_url(url):
```
Comment (rpitonak) [inline ogr/parsing.py]: missing return type

File: ogr/parsing.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    """http://github.com/foo/bar.git -> bar"""
+    repo_url = parse_git_repo(url)
+    if repo_url:
+        return repo_url.repo
+    return None
+
+
+def strip_dot_git(url):
```
Comment (rpitonak) [inline ogr/parsing.py]: missing return type

File: ogr/factory.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+from typing import Dict, Type, List
+
+from ogr.abstract import GitService, GitProject
+from ogr.exceptions import OgrException
+from ogr.parsing import parse_git_repo
+
+_SERVICE_MAPPING: Dict[str, Type[GitService]] = {}
+
+
+def use_for_service(service, _func=None):
```
Comment (rpitonak) [inline ogr/factory.py]: I would like to see a docstring here, as decorators could be hard to understand.

File: tests/unit/test_parsing.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                scheme="git+https",
+                hostname="host.name",
+            ),
+        ),
+        (
+            "git@host.name:namespace/repo",
+            RepoUrl(
+                repo="repo",
+                namespace="namespace",
+                scheme="http",
```
Comment (jpopelka) [inline tests/unit/test_parsing.py]: is this ok?

Comment (lachmanfrantisek) [reply tests/unit/test_parsing.py]: I was not sure about that either, but leave that as it was originally in Betka.

For the purpose of OGR, we always need the `http`/`https` for calling the APIs. What about using `git` and force the `http`/`https` in the service classes?

Comment (TomasTomecek) [reply tests/unit/test_parsing.py]: `git` = `http`, can we enforce https?

**Comment by lachmanfrantisek** [comment]
> (that diff is scary)

That's one downside of the current testing setup -- the real changes are hidden in the yaml diff mess.

**Comment by lachmanfrantisek** [comment]
I am going to merge this PR and fix the potential problems in the follow-up PR if needed.. I need to built another PR on top of this one.

**Pull request merged**